include/fuse_lowlevel.h

```diff
@@ -1968,27 +1970,37 @@ int fuse_session_mount(struct fuse_session *se, const char *mountpoint);
 int fuse_session_loop(struct fuse_session *se);
 
 #if FUSE_USE_VERSION < 32
```

Comment:
I think this is not necessary. If you recompile a program from source, then we can always use the new struct fuse_loop_config (since all code that works with it will be recompiled to take into account the new size). We just have to make sure that the additional member is properly initialized.

The backwards compatibility problem occurs when an existing binary uses a newer version of the libfuse shared object. In that case, fuse_lowlevel.h is never evaluated, so you can't handle this through use of FUSE_USE_VERSION. This will instead be handled through the symver declarations that you already added below.

Reply:
So lets assume we keep "struct fuse_loop_config" with the additional member and now a program initializes members of that struct without using the constructor and setter - how do we ensure without these changes here that any new struct members are initialized?
I'm more tempted to make the condition for UCLIBC and APPLE more generic (with a bit code dup), to make it better readable.

Reply:
Hmm, right, we could end up with an undefined value in `max_threads`.

How about this though? Since use of `max_idle_threads` is discouraged, re-purpose it as a flag value. If `max_idle_threads == INT_MAX`, assume that `max_threads` is defined and use it and ignore `max_idle_threads`. If `max_idle_threads` has any other value, assume that `max_thread` is undefined and use some suitable default while continuing to honor `max_idle_threads`.

Reply:
We would still need to expose the struct, instead of having it private. And with that there is no way to enforce users to set whatever value.
Let me try something else, I _think_ we can keep the old name with a more ifdefs on the version. I'm right now on kernel thread changes, will take care of this patch here a bit later today or tomorrow.
Thanks for your reviews and discussion!

Reply:
While ideally the struct would be private, I do not consider it worth the extra headache that it causes here. If I'm not mistaken, the above approach means no need for `#ifdefs` and no need for `symver` declarations. This makes the code much easier to understand and maintain.

We can still provide accessor functions and with an opaque pointer type, and recommend to use this interface, without hiding the struct completely.

If someone really wants to mess with internals, then making the struct private is not going to stop them either. They can look at the fuse source code, copy over the private definition, and still mess with it :-).

Reply:
Well, all the headache is because the struct was not private before. 

I also don't mean that user mess on purpose with internals Someone not carefully looking at the code (that could be easily me at 5 am in the morning after a long long day) might miss to initialize one or multiple values. That is basically were relying on flag values does not help.

And the other thing is, once you make it public (and allow manual settings with that) - any new extension of the struct means that a new FUSE_SYMVER is needed for each and every new parameter - really inconvenient from my point of view. Example, I will probably need another option for my current kernel changes - we really want to have another API definition for that? I'm also trying to get time (for me or a colleague) to work on ioctl threads with shared memory access - similar to zufs and fuse2. More options - another FUSE_SYMVERS needed?

Reply:
Also, please consider that the fuse version further increases for other reasons. Now file systems increase their version for those other reasons. And without noticing the file system passes an uninitialized struct.
From my point of view passing a struct that can change at any time a severe API bug. I will split the patches once we come to a conclusion how to handle this - the API bug needs a separate commit message. 
From my point of view we are left with 3 solutions
1) As in my patch, introduce a new private struct
2) Try to handle the public < 311 struct with ifdefs. From what I have in my mind it might work, but I have not written the code yet.
3) Give up on the struct an go back to the scheme that passes all parameters to the function directly.

All these 3 schemes enforce the right usage by giving compilation errors.

Reply:
I agree that having a private struct is better design, and we should have done this from the get go. However, we did not. We are having a public struct, and my worry is that replacing it with a better API will just mean that a large number of filesystems will never migrate to the new API because it's too much trouble. It would be nice if we could offer these filesystems a "minimum effort" route to get rid of the thread creation overhead.

So, in other words, I am not opposed to adding a proper API, I would just prefer if we could also offer another, easier way for filesystems to avoid the thread creation/destruction overhead.

Reply:
As i wrote, your proposal using flags does not work once the fuse version is increased for other reasons and then file systems do set the new version for that reason - the file system developer will not get automatically informed that there is an API issue. With the hundeds or thousands of different fuse file systems out there this will cause accidental read beyond end of struct - I would be scared of security CVEs...
And as I wrote, increasing the API version for every new option is not good either. 

Now, regarding thread creation overhead - from my point of view we could just ignore that idle parameter and document that it gets ignored in 3.11 . Or we add another parameter and the parameter is only not ignored if that new parameter is given. I personal do not see where thread destruction is needed, but who knows. For my upcoming kernel patches I will need at least one stable thread that will never get destructed... zufs and the unmaintained fuse2 branch from Miklos require to create a thread per core...

Reply:
There another way - security-by-obscurity. 

```
#define CONFIG_VERSION_VERIFIER1 <some-big-random-number>
#define CONFIG_VERSION_VERIFIER2 <another-big-random-number>
#define CONFIG_VERSION_VERIFIER_STR "I-know-that-I-shall-not-use-this-struct-muanlly-and-I-know-might-cause-read-of-uninitialized-variables-and-security-issues-by-doing-so"

/**
 * beginning with  FUSE_MAKE_VERSION(3, 11)  this struct should be treated a private/opaque
 * and only be accessed through fuse_loop_config_create() and fuse_loop_config_set_*() setters
 * outside of libfuse internals.
 */
struct fuse_loop_config {
...
    unsigned int max_threads;
    
    /* FUSE_MAKE_VERSION(3, 11) starts below  */
    
    uint64_t verifier1;
    uint64_t verifier2;
    char      verifier3[sizeof(CONFIG_VERSION_VERIFIER_STR )+ 1];
}
```

And with that  fuse_loop_config_create() would set these values and we would need another function like fuse_loop_config_verify() that would verify these values.

Reply:
@BenBE Thanks. This might work, although I'm not entirely sure about "code compiled with the new headers will use the new definitions.". We basically want that code compiled with FUSE_USE_VERSION < 311 gets compiled with the old definition of struct fuse_loop_config. Otherwise recompilation of a file system that was compiling fine before 3.11 would now fail - might cause a mass debian bug report (same for other distributions). And I _think_ that requirement brings in the need for #ifdefs, to define struct fuse_loop_config as the old version for that code. 
I guess the result is that a versioned struct is then not so different to an internal private struct, with setters and getters. Yours has the advantage that a conversion function is not needed, with the disadvantage of the need for more future versioning.

Reply:
Well, we could reserve some space into fuse_loop_config_versioned for future options, to avoid the need for more FUSE_SYMVER on each and every struct increase.

Reply:
Okay, didn't see that requirement for explicit targeting of old header versions. though could possibly be accomplished by overlaying `fuse_loop_config` as a `#define fuse_loop_config fuse_loop_config_v1`. Absolutely not a situation with a good solution.

Possibly the least messiest idea is to just drop providing `fuse_loop_config` as a struct type itself, but just `fuse_loop_config_versioned`, `fuse_loop_config_v1`, `fuse_loop_config_v2`. Depending on targeted API version `fuse_loop_config` could map to `fuse_loop_config_v1` for pre-3.11 and `fuse_loop_config_versioned` after. You could choose to just embed extra fields directly in `fuse_loop_config_versioned`, but at least for the two fields of `fuse_loop_config_v1` you want an union so typecasting/copying the data between the legacy version and the new one stays straight-forward.

Re conversion:
On entry to any function you basically have:

```c
struct fuse_loop_config_v1 *cfg_unk = (struct fuse_loop_config_v1*)cfg;
const bool have_v1 = cfg_unk->clone_fd != magic;
struct fuse_loop_config_versioned cfg_v = have_v1 ? (struct fuse_loop_config_versioned){ .version = magic, .v1 = *cfg_unk, .v2 = default_v2, … } : *(struct fuse_loop_config_versioned*)cfg;
```

To pass this on you'd do:
```c
#define pass_on(cfg) ((struct fuse_loop_config*)(have_v1 ? &cfg.v1 : &cfg))
```

Both could be further optimized to hide some gory details, but starting from there it shouldn't introduce too much complexity apart from this versioning transition check, which can mostly be hidden when working with things internally. You might even get away with just rerouting the passed config pointer only when you receive a _v1 structure and otherwise using things as is.

Reply:
Yeah that works, but we still need to have reserved fields in the struct for future extensions (to avoid access beyond passed memory) to avoid FUSE_SYMVER() for every new config  option. So why shouldn't we do the special work only for FUSE_USE_VERSION < 311 and from 311 on we can use the API as in my initial patch? Well, I'm fine with anything as long as I can move ahead - my other patches are depending on this patch :)
The discussion with @Nikratio happend, because Nikolaus disagreed on adding an struct ifdef for < 311. I think your suggestions are more complicated that that macro. Well, I need to upload a new patch to let you have look.

Reply:
Just uploaded an updated version. Basically the old struct is available in fuse_common.h

```
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3, 11)
struct fuse_loop_config_v1; /* forward declarition */
struct fuse_loop_config {
#else
struct fuse_loop_config_v1 {
#endif
	/**
```

and the new struct is in fuse_i.h

```
#if FUSE_USE_VERSION < FUSE_MAKE_VERSION(3,11)
struct fuse_loop_config_v2
#else
struct fuse_loop_config
#endif
{
```

What do you think?